### PR TITLE
G-API: Removing ParseSSD overload.

### PR DESCRIPTION
**Pull request by mpashchenkov**
### parseSSD overload is ambiguous.

Added specialization for ParseSSD functions with different return values. Unfortunately needs call `cv::gapi::parseSSD<GArray<Rect>>` for second version of ParseSSD. But all samples use first version of ParseSSD with tuple in return value. 

Returned overloads, removed defaults from second function variant.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Magic commands:
```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp [hunk 1]
```diff
@@ -52,7 +52,7 @@ namespace parsers {
 /** @brief Parses output of SSD network.
 
 Extracts detection information (box, confidence, label) from SSD output and
-filters it by given confidence and label.
+filters it by given confidence and label. FBL - filter by label.
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/parsers.hpp:55]: According to the `OOB` it can be just `BL`

File: modules/gapi/src/backends/cpu/gcpucore.cpp [hunk 2]
```diff
@@ -598,7 +598,7 @@ GAPI_OCV_KERNEL(GCPUParseSSDBL, cv::gapi::nn::parsers::GParseSSDBL)
     }
 };
 
-GAPI_OCV_KERNEL(GOCVParseSSD, cv::gapi::nn::parsers::GParseSSD)
+GAPI_OCV_KERNEL(GCPUParseSSD, cv::gapi::nn::parsers::GParseSSD)
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/cpu/gcpucore.cpp:601]: Why `CPU` ? I'm ok with `OCV`

**Review by TolyaTalamanov — COMMENTED** [review]
`parseFBL` and `parseSSDOOB` look confusing. Suggest to keep both of them as a single operation `parseSSD` and it have to have single implementation as well. 

From my point of view need to redesign input parameters.

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp [hunk 3]
```diff
@@ -85,11 +86,11 @@ the larger side of the rectangle.
...
-GAPI_EXPORTS GArray<Rect> parseSSD(const GMat& in,
-                                   const GOpaque<Size>& inSz,
-                                   const float confidenceThreshold = 0.5f,
-                                   const bool alignmentToSquare = false,
-                                   const bool filterOutOfBounds = false);
+GAPI_EXPORTS GArray<Rect> parseSSDOOB(const GMat& in,
+                                      const GOpaque<Size>& inSz,
+                                      const float confidenceThreshold = 0.5f,
+                                      const bool alignmentToSquare = false,
+                                      const bool filterOutOfBounds = false);
```
Comment (dkurt) [inline modules/gapi/include/opencv2/gapi/infer/parsers.hpp:93]: Is there a way to keep a single method by removing some arguments? As for me, `filterOutOfBounds` is a redundant one. OOF boxes should be filtered anyway - there is no point to provide them for user.

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp [hunk 4]
```diff
@@ -64,14 +64,15 @@ detection is smaller than confidence threshold, detection is rejected.
...
 @return a tuple with a vector of detected boxes and a vector of appropriate labels.
 */
-GAPI_EXPORTS std::tuple<GArray<Rect>, GArray<int>> parseSSD(const GMat& in,
-                                                            const GOpaque<Size>& inSz,
-                                                            const float confidenceThreshold = 0.5f,
-                                                            const int   filterLabel = -1);
+GAPI_EXPORTS std::tuple<GArray<Rect>, GArray<int>> parseSSDFBL(const GMat& in,
+                                                               const GOpaque<Size>& inSz,
+                                                               const float confidenceThreshold = 0.5f,
+                                                               const int   filterLabel = -1);
```
Comment (dkurt) [inline modules/gapi/include/opencv2/gapi/infer/parsers.hpp:70]: Is `filterLabel` used to filter background class id? Why not to use corresponding field from `DetectionOutput` layer? (named `background_label_id`)

**Comment by asenyaev** [comment]
jenkins cn please retry a build

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp (hunk 1 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/infer/parsers.hpp]: Added specialization instead.

File: modules/gapi/src/backends/cpu/gcpucore.cpp (hunk 2 above)
Comment (mpashchenkov) [reply modules/gapi/src/backends/cpu/gcpucore.cpp]: Inline with other.

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp (hunk 3 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/infer/parsers.hpp]: Needs save current interface for one of two versions.

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp (hunk 4 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/infer/parsers.hpp]: This parameter provides id of label (is `DetectionOutput layer`[1]) which needs to skip.

**Comment by dmatveev** [comment]
@mpashchenkov is this PR still actual or you have another solution now? Where did that clean-up happen? Should we close this one?

**Comment by mpashchenkov** [comment]
@dmatveev, this PR contains the last decision we made. It is actual.

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp [hunk 5]
```diff
@@ -9,7 +9,6 @@
 #define OPENCV_GAPI_PARSERS_HPP
 
 #include <utility> // std::tuple
-
```
Comment (TolyaTalamanov) [inline modules/gapi/include/opencv2/gapi/infer/parsers.hpp:12]: Avoid unnecessary changes

File: modules/gapi/src/backends/cpu/gnnparsers.hpp [hunk 6]
```diff
@@ -11,19 +11,14 @@
...
-                std::vector<cv::Rect>& out_boxes,
-                std::vector<int>&      out_labels);
-
-void parseSSD(const cv::Mat&  in_ssd_result,
-              const cv::Size& in_size,
-              const float     confidence_threshold,
-              const bool      alignment_to_square,
-              const bool      filter_out_of_bounds,
-              std::vector<cv::Rect>& out_boxes);
+void unitedParseSSD(const cv::Mat&  in_ssd_result,
```
Comment (TolyaTalamanov) [inline modules/gapi/src/backends/cpu/gnnparsers.hpp:14]: `parseSSD` ?

**Review by TolyaTalamanov — APPROVED** [review]

File: modules/gapi/include/opencv2/gapi/infer/parsers.hpp (hunk 5 above)
Comment (mpashchenkov) [reply modules/gapi/include/opencv2/gapi/infer/parsers.hpp:12]: Returned.

File: modules/gapi/src/backends/cpu/gnnparsers.hpp (hunk 6 above)
Comment (mpashchenkov) [reply modules/gapi/src/backends/cpu/gnnparsers.hpp]: Fixed.

**Review by TolyaTalamanov — APPROVED** [review]

**Comment by TolyaTalamanov** [comment]
@alalek Can it be merged ?
